relay-server/src/services/buffer/envelope_repository/sqlite.rs

```diff
@@ -0,0 +1,645 @@
+    pub fn new_with_store(config: &Config, envelope_store: SqliteEnvelopeStore) -> Self {
+        Self {
+            envelope_stacks: HashMap::new(),
+            envelope_store,
+            buffered_envelopes_size: 0,
+            disk_batch_size: config.spool_envelopes_stack_disk_batch_size(),
+            max_disk_size: config.spool_envelopes_max_disk_size(),
+        }
+    }
+
+    /// Initializes the envelope repository.
+    ///
+    /// Retrieves the project key pairs from the envelope store and creates
+    /// an initialization state.
+    pub async fn initialize(&mut self) -> InitializationState {
+        match self.envelope_store.project_key_pairs().await {
+            Ok(project_key_pairs) => {
+                self.initialize_empty_stacks(&project_key_pairs);
+                InitializationState::new(project_key_pairs)
+            }
+            Err(error) => {
+                relay_log::error!(
+                    error = &error as &dyn Error,
+                    "failed to initialize the sqlite stack repository"
+                );
+                InitializationState::empty()
+            }
+        }
+    }
+
+    /// Pushes an envelope to the repository for the given project key pair.
+    ///
+    /// If the spool threshold is exceeded, it may trigger spooling to disk.
+    pub async fn push(
+        &mut self,
+        project_key_pair: ProjectKeyPair,
+        envelope: Box<Envelope>,
+    ) -> Result<(), SqliteEnvelopeRepositoryError> {
+        if self.above_spool_threshold() {
+            self.spool_to_disk().await?;
+        }
+
+        self.envelope_stacks
+            .entry(project_key_pair)
+            .or_default()
+            .cached_envelopes
+            .push(envelope);
+
+        self.buffered_envelopes_size += 1;
+
+        Ok(())
+    }
+
+    /// Peeks at the next envelope for the given project key pair without removing it.
+    ///
+    /// If no envelope is in the buffer, it will be loaded from disk and a reference will be
+    /// returned.
+    pub async fn peek(
+        &mut self,
+        project_key_pair: ProjectKeyPair,
+    ) -> Result<Option<&Envelope>, SqliteEnvelopeRepositoryError> {
+        // If we have no data for the project key pair, we can safely assume we don't have envelopes
+        // for this pair anywhere.
+        let Some(envelope_stack) = self.envelope_stacks.get(&project_key_pair) else {
+            return Ok(None);
+        };
+
+        if envelope_stack.cached_envelopes.is_empty() && envelope_stack.check_disk {
+            let envelopes = self.unspool_from_disk(project_key_pair, 1).await?;
+            // If we have no envelopes in the buffer and no on disk, we can be safe removing the entry
+            // in the buffer.
+            if envelopes.is_empty() {
+                self.envelope_stacks.remove(&project_key_pair);
+                return Ok(None);
+            }
+
+            self.buffered_envelopes_size += envelopes.len() as u64;
```

Comment:
nit: This got me confused because for a minute I thought `buffered_envelopes_size` counted the total number of envelopes in the repository. Should we rename it to `cached_envelopes_size`?